active/0020-node-evacuation.md

```diff
@@ -0,0 +1,356 @@
+    to some other node and takeover their sessions.
+* Node evacuation with enforced session takeover. In this variant, we additionally start takeover processes
+for the sessions kept on the node. Session data is preserved on some other node even if a client does not reconnect itself
+to a new node.
+
+To implement this logic we introduce two major components:
+* _Eviction Agent_
+* _Evacuation Coordinator_
+
+_Eviction Agent_ is an entity that:
+* Blocks incoming connection on its node.
+* Serves as a facade for connection/session eviction operations.
+* Reports blocking status via HTTP API.
+
+Eviction Agent is quite passive, it does not have any complex eviction logic.
+
+_Evacuation Coordinator_ is an entity that
+* Enables Eviction Agent on the local node.
+* Sends connection/session eviction commands to it according to the evacuation logic.
+
+Evacuation Coordinator contains logic for active evacuation but knows little about connection/session nature.
+
+This component decomposition is chosen to decouple working with low-level details (hooks, connections, sessions, connection managers, ...) from the high-level logic of evacuation.
+
+That should allow us to reuse Eviction Agents in more complicated scenarios, like node rebalancing: we may implement
+some `Rebalance Coordinator` that
+* enables Eviction Agents on some nodes;
+* issues connection/session eviction commands to achieve some desired state of connection distribution.
+
+## Components
+
+### Virtual Connection pool
+
+To preserve sessions for clients that do not reconnect we implement a _virtual_
+connection. It is a simplified version of `emqx_connection`/`emqx_ws_connection`
+that does not have a real associated socket or WS connection.
+
+These virtual connections do not run under `esockd` or `ranch` supervisors but under
+their own supervisor.
+
+### Eviction Agent
+
+Eviction Agent is a stateful process and a collection of supporting modules. Eviction Agent is
+passive and responsible for disconnecting clients.
+It includes:
+* Stateful process that provides eviction API and incapsulates eviction state.
+* HTTP API reporting current status and providing endpoints for health checks.
+
+Eviction Agent states are the following:
+
+![Eviction Agent Statuses](0020-assets/eviction-agent-statuses.png)
+
+#### Internal API
+
+Eviction Agent exposes API for the following actions:
+* In disabled state:
+    * enable permanently or temporarily;
+    * provide stats abount connected/disconnected client count;
+    * provide stats abount session count.
+* In either of enabled states:
+    * disable;
+    * provide stats abount connected/disconnected client count;
+    * provide stats abount session count;
+    * disconnect certain amount of connected clients;
+    * evict certain amount of sessions.
+
+#### HTTP API
+
+Eviction Agent exposes two endpoints: `/node_eviction/status` and `/node_eviction/heath_check`.
```

Comment:
for health check, maybe extend the `/status` API.
maybe change `node_eviction` to `rebalance`